packages/contracts-bedrock/test/L2/L2ToL2CrossDomainMessenger.t.sol

```diff
@@ -0,0 +1,731 @@
+
+        // Look for correct emitted event
+        vm.expectEmit(Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER);
+        emit L2ToL2CrossDomainMessenger.RelayedMessage(
+            keccak256(abi.encode(block.chainid, _source, _nonce, _sender, target, message))
+        );
+
+        // Ensure the target contract is called with the correct parameters
+        vm.expectCall({ callee: target, msgValue: _value, data: message });
+
+        // Ensure caller is CrossL2Inbox to prevent a revert from the caller check and that it has sufficient value
+        hoax(Predeploys.CROSS_L2_INBOX, _value);
+
+        // Call the relayMessage function
+        l2ToL2CrossDomainMessenger.relayMessage{ value: _value }({
+            _destination: block.chainid, // ensure the destination is the chain of L2ToL2CrossDomainMessenger
+            _source: _source,
+            _nonce: _nonce,
+            _sender: _sender,
+            _target: target,
+            _message: message
+        });
+
+        // Check that successfulMessages mapping updates the message hash correctly
+        assertEq(
+            l2ToL2CrossDomainMessenger.successfulMessages(
+                keccak256(abi.encode(block.chainid, _source, _nonce, _sender, target, message))
+            ),
+            true
+        );
+
+        // Check that entered slot is cleared after the function call
+        assertEq(l2ToL2CrossDomainMessenger.entered(), false);
+
+        // Check that metadata is cleared after the function call. We need to set the `entered` slot to non-zero value
+        // to prevent NotEntered revert when calling the crossDomainMessageSender and crossDomainMessageSource functions
+        L2ToL2CrossDomainMessengerWithModifiableTransientStorage(Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER)
+            .setEnteredValueTrue();
+        assertEq(l2ToL2CrossDomainMessenger.crossDomainMessageSource(), 0);
+        assertEq(l2ToL2CrossDomainMessenger.crossDomainMessageSender(), address(0));
+    }
+
+    /// @dev Mock reentrant function that calls the `relayMessage` function.
+    /// @param _source Source chain ID of the message.
+    /// @param _nonce Nonce of the message.
+    /// @param _sender Sender of the message.
+    function mockTargetReentrant(uint256 _source, uint256 _nonce, address _sender) external payable {
+        // Ensure caller is CrossL2Inbox to prevent a revert from the caller check
+        vm.prank(Predeploys.CROSS_L2_INBOX);
+
+        // Ensure that the contract is entered
+        assertEq(l2ToL2CrossDomainMessenger.entered(), true);
+
+        vm.expectRevert(ReentrantCall.selector);
+
+        l2ToL2CrossDomainMessenger.relayMessage({
+            _destination: block.chainid,
+            _source: _source,
+            _nonce: _nonce,
+            _sender: _sender,
+            _target: address(0),
+            _message: ""
+        });
+
+        // Ensure the function still reverts if `expectRevert` succeeds
+        revert();
```

Comment:
I am confused about this revert

Reply:
basically, the `expectRevert()` a few lines above prevents the revert from the reentrant call from making it to initial call, so the initial call ends up emitting `RelayedMessage` instead of `FailedRelayedMessage`. the line `revert()` in the reentrant calls ensures that if the `expectRevert()` is successful (i.e., reentrancy failed), the initial call still emits `FailedRelayedMessage` and not the event for a successful message